app/scripts/controllers/user-storage/encryption.ts

```diff
@@ -75,33 +143,47 @@ class EncryptorDecryptor {
 
     const encryptedPayload: EncryptedPayload = {
       v: '1',
+      t: 'scrypt',
       d: encryptedData,
-      iterations: this.#PBKDF2_ITERATIONS,
+      o: {
+        N: this.#SCRYPT_N,
+        r: this.#SCRYPT_r,
+        p: this.#SCRYPT_p,
+        dkLen: this.#ALGORITHM_KEY_SIZE,
+      },
     };
 
     return JSON.stringify(encryptedPayload);
   }
 
   #decryptStringV1(data: EncryptedPayload, password: string): string {
-    const { iterations, d: base64CiphertextAndNonceAndSalt } = data;
+    const { o, d: base64CiphertextAndNonceAndSalt } = data;
 
     // Decode the base64.
     const ciphertextAndNonceAndSalt = base64ToByteArray(
       base64CiphertextAndNonceAndSalt,
     );
 
     // Create buffers of salt and ciphertextAndNonce.
-    const salt = ciphertextAndNonceAndSalt.slice(0, this.#PBKDF2_SALT_SIZE);
+    const salt = ciphertextAndNonceAndSalt.slice(0, this.#SCRYPT_SALT_SIZE);
     const ciphertextAndNonce = ciphertextAndNonceAndSalt.slice(
-      this.#PBKDF2_SALT_SIZE,
+      this.#SCRYPT_SALT_SIZE,
       ciphertextAndNonceAndSalt.length,
     );
 
-    // Derive the key using PBKDF2.
-    const key = pbkdf2(sha256, password, salt, {
-      c: iterations,
-      dkLen: this.#ALGORITHM_KEY_SIZE,
-    });
+    // Derive the key.
+    const cachedKey = getCachedKeyBySalt(salt);
+    let key: Uint8Array;
+    if (cachedKey) {
+      key = cachedKey.key;
+    } else {
+      key = scrypt(password, salt, {
```

Comment:
(nitpick) the logic of checking the cache result and calling scrypt is unnecessarily duplicated.

Hiding the actual key derivation behind a more generic function would make replacing/reconfiguring it be a smaller change in the future.

Reply:
Good callout!